### DISCUSSION TITLE
Automatic selection of platform backend

### DISCUSSION THREAD

**Discussion by philn** [Igalia/cog#259] [Q&A] [answered]
I was wondering if this would make sense:

If a wayland compositor is running and supports SSD, use FDO backend. If SSD is not supported, try the GTK4 backend, if available. If no compositor is running, try the DRM backend. Otherwise if Xorg is running, use the X11 backend.

If a `-P` option is provided, use that backend.

WDYT? @aperezdc ?

**Comment by aperezdc** [comment]
Yes, this makes total sense! As a matter of fact, in the [devel/multiple-views](https://github.com/Igalia/cog/commits/devel/multiple-views) branch I have automatic selection implemented: each module implements a [GIO extension point](https://developer.gnome.org/gio/stable/gio-Extension-Points.html) and one of the functions in the interface allows querying the module whether it's usable in the current environment, and each module has a different priority, so they are probed in order until one of them claims that it can be instantiated.

The main commit implementing that logic is https://github.com/Igalia/cog/commit/d5243b577c5f80192e7aff1943ce60b4fcce6cb2 and I am planning to port it to the `master` branch in the coming weeks after creating teh 2.10.x release branch.

**Comment by aperezdc** [comment]
The “compositor is running” part is easy to check (try connecting) but the “and supports SSD” part is trickier because a compositor can support SSD, the client can request SSD, and yet the compositor may fail to comply with the requeast and tell the client to use CSD (e.g. the compositor may let the user configure which applications get to use SSD or CSD!). In practice, a Wayland client _should_ be ready to use CSD regardless of what the compositor claims to support. 

We might want to use something like [libdecoration](https://gitlab.gnome.org/jadahl/libdecoration) instead :thinking:

**Reply by philn** [reply]
Ah ok so that means our FDO platform plugin should also support CSD?

**Reply by aperezdc** [reply]
That is my understanding, yes.

**Comment by aperezdc** [comment, marked as answer]
FWIW, I have implemented automatic selection of platform plug-ins in PR https://github.com/Igalia/cog/pull/320 😉
